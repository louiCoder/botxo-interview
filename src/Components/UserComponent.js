import React from 'react';
import './styles.scss';

export default function UserComponent ({ changeName, changeAge }) {
  return (
    <div className="users-container">
      <div className="name-and-age">
        <div>
          <p>Name</p>
          <input type="text" onChange={(e) => changeName(e)} className="input-field" />
        </div>
        <div>
          <p>Age</p>
          <input type="number" onChange={(e) => changeAge(e)} className="input-age" />
        </div>
      </div>
      {/* <div>
        <img src={Trash} alt="delete_icon" onClick={() => alert()} />
        <svg style={{ width: '24px', height: '24px' }} viewBox="0 0 24 24" onClick={() => alert()}>
          <path
            fill="currentColor"
            d="M19,4H15.5L14.5,3H9.5L8.5,4H5V6H19M6,19A2,2 0 0,0 8,21H16A2,2 0 0,0 18,19V7H6V19Z"
          />
        </svg>
      </div> */}
    </div>
  );
}
