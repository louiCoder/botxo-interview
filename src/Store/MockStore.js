import configureMockStore from 'redux-mock-store';
import sagaMiddleware from 'redux-saga';
import reducer from './Reducer';

const middlewares = [ sagaMiddleware ];

export default configureMockStore(reducer, middlewares);
