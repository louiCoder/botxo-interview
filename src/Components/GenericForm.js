import React, { useState } from 'react';
import './styles.scss';
import UserComponent from './UserComponent';
import { useDispatch } from 'react-redux';
import { addGenericUser } from '../Store/ActionCreators';

export default function GenericForm () {
  const [ name, setName ] = useState('');
  const [ age, setAge ] = useState(0);
  const dispatch = useDispatch();

  const addToGenericList = () => {
    if (!name || !age) return alert(`${!name ? 'name' : 'age'} is missing`);
    dispatch(addGenericUser(name, age));
  };
  return (
    <div className="generic-form-container">
      <h3>Generic Data</h3>
      <div>
        <UserComponent
          changeName={(e) => setName(e.target.value)}
          changeAge={(e) => setAge(e.target.value)}
          name={name}
          age={age}
        />
        {/* <UserComponent onChange={() => null} changeAge={() => null} /> */}
      </div>
      <button className="send-button" onClick={addToGenericList}>
        Add user in generic and specific data
      </button>
    </div>
  );
}
