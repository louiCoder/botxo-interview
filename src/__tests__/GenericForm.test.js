import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import GenericForm from '../Components/GenericForm';
import UserComponent from '../Components/UserComponent';
import MockStore from '../Store/MockStore';

describe('<GenericForm /> component', () => {
  let store, wrapper, props;

  beforeEach(() => {
    store = MockStore({ genericUsers: [], specificUsers: [], loading: false });
    props = {
      changeName: jest.fn(),
      setName: jest.fn(),
      addToGenericList: jest.fn(),
      name: '',
      age: 0,
      setAge: jest.fn()
    };
    wrapper = mount(
      <Provider store={store}>
        <GenericForm {...props} />
      </Provider>
    );
  });

  it('Render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call ADD_GENERIC_USER_ACTION on button click while name and age exist', () => {
    wrapper.find(UserComponent).find('input').first().simulate('change', { target: { value: 'louis' } });
    wrapper.find(UserComponent).find('input').last().simulate('change', { target: { value: 20 } });

    wrapper.find('button').simulate('click');
    expect(store.getActions()).toEqual([ { type: 'ADD_GENERIC_USER_ACTION', name: 'louis', age: 20 } ]);
  });

  it('simulate a change name event on child component <UserComponent />', () => {
    wrapper.find(UserComponent).find('input').first().simulate('change', { target: { value: 'louis' } });
    expect(wrapper.find(UserComponent).find('input').first().props().value).toEqual('louis');
  });

  it('simulate a change age event on child component <UserComponent />', () => {
    wrapper.find(UserComponent).find('input').last().simulate('change', { target: { value: 20 } });
    expect(wrapper.find(UserComponent).find('input').last().props().value).toEqual(20);
  });
});
