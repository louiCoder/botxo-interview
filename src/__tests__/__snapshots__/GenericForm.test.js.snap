// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenericForm /> component Render without crashing 1`] = `ReactWrapper {}`;
